### DISCUSSION TITLE
ST7796S with Blackpill

### DISCUSSION THREAD

**Discussion by quartersawn** [Bodmer/TFT_eSPI#3557] [Q&A - General]
I try to be "self-service" on my projects but I'm stumped on this one. I was able to successfully build a TFT_eSPI project with some sample code using a WeAct Blackpill V3 (STM32F411)  and a ST7786 display over SPI. But I needed a bigger display, so I switched to a Hosyond ST7796S link [here](https://www.amazon.com/dp/B0CRGQN58D?ref=ppx_yo2ov_dt_b_fed_asin_title&th=1). (the one I have has a touch screen, but that is not being implemented - yet).

I used the same pins that worked for the ST7786, but have not been able to get the ST7796 to work - just a backlit screen. I'm doing my configuration via platformio.ini. The relevant lines are:
        -D USER_SETUP_LOADED=1
	-D ST7796_DRIVER=1
	-D TFT_WIDTH=480
	-D TFT_HEIGHT=320
	-D TFT_MOSI=PA5
	-D TFT_SCLK=PA7
	-D TFT_CS=PB13
	-D TFT_DC=PB12
	-D TFT_RST=PB15
        -D SPI_FREQUENCY=27000000

The pins chosen are those available in the project and as stated above, these pins worked fine with the ST7786. I've checked the wiring multiple times, and I even connected my scope up to verify activity. (I'm very much a novice using the scope)

I'm using an Arduino framework for my projects.

I've also run the setup_test code and see this for the pins:
MOSI    = GPIO -59 
SCK     = GPIO -57
TFT_CS   = GPIO 28 //maps to PB13 as in platformio.ini
TFT_DC   = GPIO 27 //maps to PB12 as in platformio.ini
TFT_RST  = GPIO 30  //maps to PB15 as in platformio.ini

I confess that the MOSI and SCK look strange, but they work for the ST7786. Everything else in the setup_test output looks correct - driver, SPI frequency, etc.

I'm fairly sure I'm not seeing the forest for the trees, and there's a head slap coming my way, butif anyone can push me in the right direction, I'd appreciate it.

Steve

**Comment by Basitadam** [comment]
You sure that the backlight is on? I don't see LED pin configured. Is it connected directly to Vcc? Hosyond  site says leave it not connected if no control needed but make sure it is not tied to GND.

I found some differences in init sequence in the Hosyond samples. I can write a different ST7796_init.h file so you can test. Give me some time (now+12 hours), tomorrow morning, I will send a link.

Hopefully you won't need a slap

**Comment by quartersawn** [comment]
Yes, I'm sure. I read the same thing - leave it floating or connect to +5 or +3.3.

Thanks for the help with the init. I'll take a look at it in the meanwhile and see if I can make any sense of it.

**Reply by Basitadam** [reply]
Can you test with this init. Please make a copy of the original  ST7796_Init.h file first. I don't remember platformio file structure but the file is in ..\TFT_eSPI\TFT_Drivers folder.

Here is the download link to modified the file : https://www.filemail.com/d/xcxcdtykjqwtqgh

**Reply by quartersawn** [reply]
Thanks for the effort. I swapped it out and had no success. Sometimes the solution is right in front of you but you don't see it. I certainly didn't until after I tried your init with no success. Turns out the answer was in front of me all along, staring at my from the display on my scope. Check the screen grab:

![RigolDS5](https://github.com/user-attachments/assets/b5319a33-9d0d-4b60-b816-e1ad7a711ef5)

See it? 

I didn't until this morning, and I should have. 

The MOSI and SCK lines are reversed! I went back and checked again, and the display was wired correctly, the pins were correctly defined (I think), but the signals were wrong. I swapped the jumpers on the breadboard and saw a flash on the display. Reloaded the code and I have output from TFT_Print_Test.ino - at least until something happens and the display stops updating. 

But how did this happen? I have looked again at my platformio.ini, and if I reverse the configuration for MOSI and SCK in the file and  rebuild the code and check the output, **_nothing changes_**. The signals stay on the same pins. I've looked at user_setup.h and all pin assignments are commented out. I suspect I lack sufficient understanding of the setup portion, and need to do some more reading. 

I've also observed that while the display will start to run the demo TFT_Print_Test.ino, after a few loops it will stop on a solid colored screen. More to do here as well.

**Reply by Basitadam** [reply]
Wow. Why so much reset pulses? Looks like glitches. like 10ns width? They are in also SCK (meant to be MOSI) line, too. Could be false reading but there are non in D11 signal

**Reply by quartersawn** [reply]
Yes, there do seem to be a lot of resets,I believe those are glitches from my home made logic probe. (soon to be reworked) At this point, it appears from some pages I have read that the Blackpill board (STM32F411) hardware SPI pins cannot be reassigned, except perhaps by going into a bit of low level code I don't want to mess with.  Looking at my test with the ST7786, this seems to bear that out, as it is also working with SCK on pin PA5, and MOSI on PA7. This would explain why changes in platformio.ini have no effect.
